lib/utils/parse/ast.go

```diff
@@ -0,0 +1,493 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parse
+
+import (
+	"fmt"
+	"net/mail"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// Expr is a node in the AST.
+type Expr interface {
+	// Kind indicates the expression kind.
+	Kind() reflect.Kind
+	// Evaluate evaluates the expression given the evaluation context.
+	Evaluate(ctx EvaluateContext) (any, error)
+}
+
+// EvaluateContext is the evaluation context.
+type EvaluateContext struct {
+	// VarValue returns a list of values that a variable has.
+	VarValue func(VarExpr) ([]string, error)
+	// MatcherInput is the input to matchers.
+	MatcherInput string
+}
+
+// StringLitExpr encodes a string literal expression.
+type StringLitExpr struct {
+	value string
+}
+
+// VarExpr encodes a variable expression with the form "namespace.name".
+type VarExpr struct {
+	namespace string
+	name      string
+}
+
+// EmailLocalExpr encodes an expression with the form "email.local(expr)".
+type EmailLocalExpr struct {
+	email Expr
+}
+
+// RegexpReplaceExpr encodes an expression with the form "regexp.replace(expr, string, string)".
+type RegexpReplaceExpr struct {
+	source      Expr
+	re          *regexp.Regexp
+	replacement string
+}
+
+// RegexpMatchExpr encodes an expression with the form "regexp.match(string)".
+type RegexpMatchExpr struct {
```

Comment:
fyi: when doing work on predicate I investigated such a detailed node approach. my conclusions were that this very quickly became extremely cumbersome to handle a gigantic amount of node types. have you considered instead encoding these as generic function calls instead and supplying `regexp.match` as a closure instead? from my experience, that reduces the sheer code size by a lot.